### PR TITLE
Support telegram's channel file download

### PR DESCRIPTION
**Pull request by jianyun8023**
Support for downloading book from the telegram channel.

### TODO list

- [x] Implement telegram user login with account password.
- [x] Implement to get channel message using channel name.
- [x] Support telegram API frequency control, need to increase the wait when each minute call exceeds the limit
- [x] Support parallel download.

### CODE REVIEW

File: cmd/telegram.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	Run: func(cmd *cobra.Command, args []string) {
+		// Validate config
+		spider.ValidateDownloadConfig(d)
+
+		// Create the downloader
+		downloader := telegram.NewDownloader(d)
+
+		err := downloader.Exec()
+		if err != nil {
+			panic(err)
```
Comment (syhily) [inline cmd/telegram.go]: Using `log.Fatal(err)` instead.

File: telegram/auth.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
+package telegram
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"github.com/gotd/td/telegram/auth"
```
Comment (syhily) [inline telegram/auth.go:11]: These imports should be formatted by `goimport`.

File: telegram/auth.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+func (c noSignUp) AcceptTermsOfService(ctx context.Context, tos tg.HelpTermsOfService) error {
+	return &auth.SignUpRequired{TermsOfService: tos}
+}
+
+// termAuth implements authentication via terminal.
+type TermAuth struct {
+	noSignUp
+
```
Comment (syhily) [inline telegram/auth.go]: No blank line here.

File: telegram/auth.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	phone string
+}
+
+func (a TermAuth) Phone(_ context.Context) (string, error) {
```
Comment (syhily) [inline telegram/auth.go:33]: We don't need this argument, why do we add it? same below.

File: telegram/telegram.go [hunk 5]
```diff
@@ -0,0 +1,412 @@
+package telegram
+
+import (
+	"context"
+	"errors"
+	"github.com/bibliolater/bookhunter/pkg/progress"
```
Comment (syhily) [inline telegram/telegram.go]: reimports

File: telegram/telegram.go [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+	id           int
+	filename     string
+	format       string
+	fileSize     int
+	documentFile *tg.InputDocumentFileLocation
+	filePath     string
+}
+
+func NewDownloader(config *spider.Config) *downloader {
+
```
Comment (syhily) [inline telegram/telegram.go]: Remove the blank line.

File: telegram/telegram.go [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+	LastId = last
+	// Create the channel dir.
+	saveDir := path.Join(d.config.DownloadPath, rename.EscapeFilename(title))
+	if !IsDir(saveDir) {
+		if err := os.MkdirAll(saveDir, os.ModePerm); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	//err = query.Messages(d.client.API()).Search(&tg.InputPeerChannel{
```
Comment (syhily) [inline telegram/telegram.go]: Remove the commented code.

File: telegram/telegram.go [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+			//err := d.downloadFile(entity)
+			//if err != nil {
+			//	return err
+			//}
+			d.saveCurrentBookId(entity.id, last)
+			ch <- *entity
+		}
+	}
+
+	// Return to close client connection and free up resources.
```
Comment (syhily) [inline telegram/telegram.go]: Unfinished here?

File: telegram/telegram.go [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+
+	for i := range jobs {
+		jobs[i].Index = i
+		if i == 0 {
+			jobs[i].Offset = start
+		} else {
+			jobs[i].Offset = jobs[i-1].Offset + jobs[i-1].Limit
+		}
+		jobs[i].Limit = step
+		//log.Infof("part %d --> %d", jobs[i].Offset, jobs[i].Offset+jobs[i].Limit)
```
Comment (syhily) [inline telegram/telegram.go]: We can add debug level for verbose log.

**Review by syhily — CHANGES_REQUESTED** [review]
Well done for this job. A lot of code styles issues but easy to be fixed.

File: telegram/auth.go (hunk 4 above)
Comment (jianyun8023) [reply telegram/auth.go:33]: This function user tg's login support

File: telegram/telegram.go (hunk 8 above)
Comment (jianyun8023) [reply telegram/telegram.go]: This is a wrong note, remove it.

File: telegram/telegram.go (hunk 9 above)
Comment (jianyun8023) [reply telegram/telegram.go]: I'll add more verbose logs when you've finished the groundwork

**Comment by syhily** [comment]
LGTM

**Pull request merged**